### PR TITLE
Full Auto Gun Mode Component

### PR DESCRIPTION
**Pull request by m-dzianishchyts**
## What Does This PR Do
Ports https://github.com/tgstation/tgstation/pull/57084 https://github.com/tgstation/tgstation/pull/67911.
Also provides some refactoring.

Here is an example of making a gun full auto:
```dm
/obj/item/gun/energy/turret/Initialize(mapload)
	AddComponent(/datum/component/automatic_fire, 0.1 SECONDS)
	return ..()
```

## Why It's Good For The Game
Convenient full auto guns. (am ded pls nerf)

## Testing
Compiled.

https://github.com/ParadiseSS13/Paradise/assets/39908528/cb2d2819-4efb-4736-bf2f-608c7bd6c871

## Changelog
NPFC (for now)

### CODE REVIEW

File: code/__DEFINES/combat_defines.dm [hunk 1]
```diff
@@ -157,3 +157,15 @@
...
+#define AUTOFIRE_STAT_IDLE (1<<0)
+/// Gun is active and in the user hands. Wait until user does a valid click.
+#define AUTOFIRE_STAT_ALERT	(1<<1)
+/// Gun is shooting.
+#define AUTOFIRE_STAT_FIRING (1<<2)
+
+#define COMSIG_AUTOFIRE_ONMOUSEDOWN "autofire_onmousedown"
+	#define COMPONENT_AUTOFIRE_ONMOUSEDOWN_BYPASS (1<<0)
+#define COMSIG_AUTOFIRE_SHOT "autofire_shot"
+	#define COMPONENT_AUTOFIRE_SHOT_SUCCESS (1<<0)
```
Comment (lewcc) [inline code/__DEFINES/combat_defines.dm:171]: Can you move these signals to signals.dm? I know it's still kind of a mess, but it's where we expect signals to be

File: code/_onclick/drag_drop.dm [hunk 2]
```diff
@@ -47,4 +47,65 @@ actually if you do something in that proc like changing user location or whateve
 to inform the game this action was expected and its fine
 */
 /atom/proc/MouseDrop_T(atom/dropping, mob/user, params) // return TRUE if you want to prevent us click the object after it
-	return
+	return FALSE
+
+/client/MouseDown(datum/object, location, control, params)
+	if(QDELETED(object)) // Yep, you can click on qdeleted things before they have time to nullspace. Fun.
+		return
```
Comment (lewcc) [inline code/_onclick/drag_drop.dm:54]: This might be worth a TM for possible perf concerns

**Review by DGamerL — CHANGES_REQUESTED** [review]
Alongside this, please also add comments on the `SIGNAL_HANDLER` as it is done in #24824

File: code/_onclick/drag_drop.dm [hunk 3]
```diff
@@ -47,4 +47,65 @@ actually if you do something in that proc like changing user location or whateve
...
+	return TRUE
+
+/client/MouseDrag(src_object, atom/over_object, src_location, turf/over_location, src_control, over_control, params)
+	var/list/modifiers = params2list(params)
+	if(!drag_start) // If we're just starting to drag
+		drag_start = world.time
+		drag_details = modifiers.Copy()
+	mouseParams = params
+	mouse_location_UID = over_location?.UID()
+	mouse_object_UID = over_object?.UID()
```
Comment (GDNgit) [inline code/_onclick/drag_drop.dm:101]:
```suggestion
	if(istype(over_location))
		mouse_location_UID = over_location.UID()
	if(istype(over_object))
		mouse_object_UID = over_object.UID()
```
Byond doesn't always pass atoms/turfs here. Welcome to the wonderful world of screen objects!

here's the """"correct"""" arguments for this function
```
MouseDrag(src_object as null|atom in usr.client,\
          over_object as null|atom in usr.client,\
          src_location as null|turf|text in usr.client,\
          over_location as null|turf|text in usr.client,\
          src_control as text, over_control as text, params as text)
```
horrifying

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/_onclick/drag_drop.dm (hunk 2 above)
Comment (GDNgit) [reply code/_onclick/drag_drop.dm:54]: I've gone and profiled this, it's acceptable. Won't even hit a second of time in total during a round

File: code/_onclick/drag_drop.dm [hunk 4]
```diff
@@ -47,4 +47,72 @@ actually if you do something in that proc like changing user location or whateve
...
+
+/obj/screen/click_catcher/IsAutoclickable()
+	return TRUE
+
+/* The actual arguments of this:
+MouseDrag(src_object as null|atom in usr.client,\
+          over_object as null|atom in usr.client,\
+          src_location as null|turf|text in usr.client,\
+          over_location as null|turf|text in usr.client,\
+          src_control as text, over_control as text, params as text)
```
Comment (GDNgit) [inline code/_onclick/drag_drop.dm:99]:
```suggestion
	over_object as null|atom in usr.client,\
	src_location as null|turf|text in usr.client,\
	over_location as null|turf|text in usr.client,\
	src_control as text, over_control as text, params as text)
```

**Review by DGamerL — APPROVED** [review]

**Comment by DGamerL** [comment]
FISI

**Pull request merged**